PAYARA-2598 Documentation for Payara Realms Identity Stores (#666)
payara/Payara-Server-Documentation: documentation/payara-server/public-api/realm-identitystores/certificate-identity-store-definition.adoc

[[certificate-identity-store]]
= Certificate Identity Store

_Since Payara Server 5.194_

> The Payara API provides a `@CertificateIdentityStoreDefinition` and `@CertificateAuthenticationMechanismDefinition` annotation that creates an identity store and authentication mechanism to authenticate the users using the certificate realm.

Reviewer: ```suggestion
The Payara API provides the `@CertificateIdentityStoreDefinition` and `@CertificateAuthenticationMechanismDefinition` annotations that create an identity store and authentication mechanism to authenticate users using the certificate realm.
```

[[usage]]
== Usage

> The Certificate realm authentication mechanism and identity store are defined through the `@CertificateAuthenticationMechanismDefinition` and `@CertificateIdentityStoreDefinition` annotation.

Reviewer: ```suggestion
The Certificate realm authentication mechanism and identity store are defined through the `@CertificateAuthenticationMechanismDefinition` and `@CertificateIdentityStoreDefinition` annotations.
```

> Specifying this in a valid place as defined by the Jakarta EE Security API will create the mechanism.

Reviewer: Possibly link to documentation?

Often this may mean that any class is a valid position.

> If the Certificate realm not found with the defined name then a new Certificate realm created otherwise existing Certificate realm instance is used to authenticate the users.

Reviewer: ```suggestion
If a Certificate realm is not found with the defined name then a new Certificate realm is created, otherwise the existing Certificate realm instance is used to authenticate the users.
```

[[usage-example]]
=== Example

The following code sample illustrates how to configure Certificate realm identity store:

[source, java]
----
@ApplicationScoped
@ApplicationPath("/rest")
@DeclareRoles({ "a", "b" })
@CertificateAuthenticationMechanismDefinition
@CertificateIdentityStoreDefinition("certificate-realm")
public class MyRestApp extends Application {
}
----

[[configuration]]
== Configuration

The Certificate realm identity store can be configured with both `@CertificateIdentityStoreDefinition` annotation attributes 

> and MicroProfile Config properties. The annotation and mp config properties has several configuration options.

Reviewer: ```suggestion
and MicroProfile Config properties. The annotation and MicroProfile Config properties have several configuration options.
```

They are detailed as shown below.

|===
| Option | MP Config property | Description | Default | Required

| `value`
|
| The name of realm.
| `certificate`
|

| `assignGroups`
| `payara.security.certificate.assignGroups`
| The users are assigned membership to these groups for the purposes of authorization decisions.
|
|

|===

> *Note :* If both annotation attribute and mp config property defined for same option 

Reviewer: ```suggestion
*Note :* If both an annotation attribute and a MicroProfile Config property are defined for the same option 
```

> then MicroProfile Config property value always take precedence over `@CertificateIdentityStoreDefinition` annotation value.

Reviewer: ```suggestion
then the MicroProfile Config property value always take precedence over the `@CertificateIdentityStoreDefinition` annotation value.
```
